testsuite/classlibrary/TestSynthDesc.sc

```diff
@@ -0,0 +1,88 @@
+TestSynthDesc : UnitTest {
+	test_SynthDesc_read_restores_metadata_in_desc_and_def {
+		var package = this.setupMetadataTests;
+		var synthdesc;
+
+		if(package.error.notNil) {
+			this.failed(thisMethod, "Error writing SynthDef: '%'".format(package.error.errorString));
+			^this.cleanupMetadataTests(package);
+		};
+
+		synthdesc = SynthDesc.read(package.path, keepDefs: true)[\test_md];
+		if(synthdesc.isNil) {
+			this.failed(thisMethod, "SynthDesc was not read from disk");
+			^this.cleanupMetadataTests(package);
+		};
+
+		this.checkMetadata(synthdesc, package.metadata, SynthDesc);
+
+		this.cleanupMetadataTests(package);
+	}
+
+	test_SynthDescLib_read_restores_metadata_in_desc_and_def {
+		var package = this.setupMetadataTests;
+		var synthdesc;
+
+		if(package.error.notNil) {
+			this.failed(thisMethod, "Error writing SynthDef: '%'".format(package.error.errorString));
+			^this.cleanupMetadataTests(package);
+		};
+
+		synthdesc = SynthDescLib.read(package.dir +/+ "*.scsyndef").global;
+		synthdesc = synthdesc[\test_md];
+		if(synthdesc.isNil) {
+			this.failed(thisMethod, "SynthDesc was not read from disk");
+			^this.cleanupMetadataTests(package);
+		};
+
+		this.checkMetadata(synthdesc, package.metadata, SynthDescLib);
+
+		this.cleanupMetadataTests(package);
+	}
+
+	setupMetadataTests {
+		var package = IdentityDictionary.new.know_(true);
+		package.dir = PathName.tmp;
+		package.path = package.dir +/+ "test_md.scsyndef";
+		package.mdPath = package.dir +/+ "test_md.txarcmeta";
+		package.metadata = (a: 1);
+		package.original = SynthDef(\test_md, {
```

Comment:
this `package` business feels brittle to me. it's not obvious at a glance what fields are being populated on this identity dict, and a typo in one of the field names wouldn't produce a compile error. since this file is testing the same functionality on two different classes, what if we just put this test suite in a separate file specifically for these two tests and then used class instance variables instead? then these functions could be `setup` and `teardown`, and you don't have to manually invoke them or do cleanup in early returns.

Reply:
This makes good sense to me. I've done that.

You might raise an eyebrow about starting the test method with `if(error.notNil)` but... here I'm hamstrung by a minor oversight in `UnitTest` itself -- `runTestMethod` does `this.setUp;` and *then* `currentMethod = method;` -- so, if I try to report the failure within `setUp` then it would be associated with the previous `currentMethod` (or perhaps a nil method). Unclear to me why it's done that way, but it looks like a bug to me (and out of scope of this PR).

Reply:
i agree it would be helpful to have the test method available during setup/teardown